example/no_tls_both.cpp

```diff
@@ -37,8 +37,10 @@ void client_proc(
             locked_cout() << "[client] Connack Return Code: "
                       << MQTT_NS::connect_return_code_to_str(connack_return_code) << std::endl;
             if (connack_return_code == MQTT_NS::connect_return_code::accepted) {
-                pid_sub1 = c->subscribe("mqtt_client_cpp/topic1", MQTT_NS::qos::at_most_once);
-                pid_sub2 = c->subscribe(
+                pid_sub1 = c->acquire_unique_packet_id();
+                pid_sub2 = c->acquire_unique_packet_id();
+                c->subscribe(pid_sub1, "mqtt_client_cpp/topic1", MQTT_NS::qos::at_most_once);
+                c->subscribe(pid_sub2,
```

Comment:
This code shouldn't be changed. `c->subscribe()` should return internally acquired packet_id.

Reply:
We discussed removing the return value entirely, so that the completion token for https://github.com/redboltz/mqtt_cpp/issues/523 can specify the return value.

If end user wants to know packet_I'd they must acquire it before calling publish/subscribe

Further, there is no reason to assume that acquiring the packet id will succeed. It may be the case now that the packet Id will be successfully required, but this is not a good assumption for more generic packet_Id implementation. See https://github.com/redboltz/mqtt_cpp/issues/546 and https://github.com/redboltz/mqtt_cpp/issues/223

With user supplied packet_id implementations, and async packet_id implementation. (And both) assumption of success is not possible.

Reply:
> We discussed removing the return value entirely, so that the completion token for #523 can specify the return value.

I disagree. Maybe we misunderstand something.
I need packet_id return version. I don't accept removing them.

If some special (I think it should be default) completion token is given, the API such as `publish()` should return packet_id.

I think that it is similar to `use_future`.
https://www.boost.org/doc/libs/1_71_0/doc/html/boost_asio/example/cpp11/futures/daytime_client.cpp

If completion token is `use_future` then `async_recv_from()` returns future.
Similarly, if completion token is something special (by default) then `publish()` returns packet_id.

I understand that we go on this way on #523.

Reply:
> Further, there is no reason to assume that acquiring the packet id will succeed. It may be the case now that the packet Id will be successfully required, but this is not a good assumption for more generic packet_Id implementation. See #546 and #223

I'm not sure what you mean. But if packet_id acquisition is failed, then exception is thrown.

Reply:
I think that we keep the original `packet_id` return version and, remove bool return version, it is acceptable solution.

Reply:
Why do you need the packet_id to be returned?

The *only* difference between
```
packet_id_t publish();
and
void publish(packet_id_t);
```

is that `packet_id_t publish();` calls `acquire_unique_packet_id()`, internally. If an end user requires the packet_id, they would be better off calling `acquire_unique_packet_id()` themselves, so that they have control over what to do on failure.

> I'm not sure what you mean. But if packet_id acquisition is failed, then exception is thrown.

This is not possible if packet_id acquisition is async.

> I think that it is similar to use_future.
>
>https://www.boost.org/doc/libs/1_71_0/doc/html/boost_asio/example/cpp11/futures/daytime_client.cpp

> If completion token is use_future then async_recv_from() returns future.
Similarly, if completion token is something special (by default) then publish() returns packet_id.

The completion token specifies the return value, in totality, unless we put in a lot of extra code in to override this behavior of the Boost.Asio CompletionToken concept.

I also think that you misunderstood how that example works.

`boost::asio::use_future` is the completion token passed to the async function `resolver.async_resolve`. The completion token is used inside `resolver.async_resolve` to determine the return type (`std::future<udp::resolver::results_type>`)

The `std::future<>` that is returned is constructed by the completion token (there's some template function that boost::asio provides to handle this automatically), and when the CompletionToken is then called with the result of the async operation completing, the CompletionToken provides the result to the `std::future<>`

As far as I can determine, this is the only way that mqtt_cpp can be used with `std::future<>`, or boost::coroutine2, or std coroutine.

Reply:
> Why do you need the packet_id to be returned?
> 
> The only difference between
>
> ```
> packet_id_t publish();
> and
> void publish(packet_id_t);
> ```
> is that packet_id_t publish(); calls acquire_unique_packet_id(), internally. If an end user requires the packet_id, they would be better off calling acquire_unique_packet_id() themselves, so that they have control over what to do on failure.

The reason is most of existing MQTT library return packet id.
See
https://github.com/mqttjs/MQTT.js/blob/master/lib/client.js#L415
https://github.com/eclipse/paho.mqtt.java/blob/master/org.eclipse.paho.client.mqttv3/src/main/java/org/eclipse/paho/client/mqttv3/IMqttAsyncClient.java#L364

And users are familiar with this kind of interface I guess.
I though that user might think that it is bother that one additional API call.
So I want to (continue to) provide convenient version.

But I start thinking that if user seriously want to manage packet_id, the user calls separated acquire/publish APIs.

We can keep to have `packet_id_t publish()` that calls sync `acquire_unique_packet_id()`. Why do you want to remove them?
I think that your goal is achieved without removing that. And just you don't call that APIs.

I will consider that again later.

Reply:
I considered. 
* The advantage of removing internally acquiring version is reduce functions we need to maintain.
* The advantages of keeping internally acquiring version are no breaking change and natural interface for other MQTT library users.

I tried to find a way that keeping internally acquiring version and minimize maintenance cost.
Finally I found the way. See https://wandbox.org/permlink/7TCPdFzcUYFG9i7h

The base implementation is acquired version that has packet_id parameter.
And providing only 6 overloads for internally acquiring version.
6 means, publish, subscribe, unsubscribe, async_publish, async_subscribe, and async_unsubscribe.

Could you add these overloads and revert example and test codes to use internally acquiring version?

Reply:
No.

I strongly disagree that any function should ever return a packet_id except `acquire_unique_packet_id()` or `register_packet_id()`.

Either the user cares about the packet_id, or they do not. If they care, it's only one additional function call. If they don't care, they don't care.

It's not possible to have the function return a packet_id for the async functions, because the packet_id acquisition process may fail (async packet_id acquisition). If we can't return a packet_id for the async case, then there's not a lot of reason to return the packet_id for the non-async case. The async / sync APIs should be as close to each other as practical.

And as discussed previously, returning library-defined data is not compatible with the Boost.Asio CompletionToken concept. I do not want to introduce a lot of additional template machinery when we're already got compile times that time out on Mac OS X / Travis, especially when it's not clear to me that that can ever work properly with Boost.Asio CompletionToken.

--------------------------------------------------------

I don't think that `mqtt_cpp::endpoint` should have anything to do with acquiring packet_ids. What I *want* to do is remove the `acquire_unique_packet_id()` and `register_packet_id()` functions from `mqtt_cpp::endpoint`, and have them be either in `mqtt_cpp::client`, or be separate from the `mqtt_cpp::endpoint` inheritance hierarchy entirely. If `acquire_unique_packet_id()` and `register_packet_id()`  were moved to `mqtt_cpp::client`, then `acquire_unique_packet_id()` and `register_packet_id()`  would also be moved to `mqtt_cpp::client`, and `mqtt_cpp::endpoint` would have no code that acquired the packet_id at all.

But doing that would be too much of a change to the API all at once, so I left the `acquire_unique_packet_id()` and `register_packet_id()`  functions where they were for this work

Reply:
I said that just call **sync** version of `acquire_unique_packet_id()`. If you want to removing packet_id acquisition mechanism, I don't agree with that yet. It's different topic. I don't find any value to async packet_id allocation mechanism. But you want to do that, I accept **add** this in the future. Not replace sync one.

This is philosophy difference. I'm not convinced. If you can't accept add packet_id return version, please fork.

Reply:
I do not understand your reply.

I'm not proposing removing or moving `acquire_unique_packet_id()` in this change.

I said that I think it should be moved, but that would be later, if ever.

> I don't find any value to async packet_id allocation mechanism. But you want to do that, I accept add this in the future. Not replace sync one.

I'm not proposing that the non-async packet_id allocation mechanism be removed or replaced at this time. If it is ever removed (which is unlikely), it would be when mqtt_cpp non-async code is removed (which you've said won't happen), so I am not proposing that non-async packet_id allocation mechanism be removed or replaced.

I am saying that if `async_acquire_unique_packet_id()` is used, then `packet_id_t async_publish(...);` cannot exist. It is not possible.

I wish to use `async_acquired_unique_packet_id()` because I store all of my program state in local a database that has an async API. non-async `acquired_unique_packet_id()` makes this *blocking* operation, which is very poor performance. 

> The advantages of keeping internally acquiring version are no breaking change

Unfortunate, but necessary. Rip off the band-aide.

> I think that user might think that it is bother that one additional API call.
> So I want to (continue to) provide convenient version.

Aside from you, who is even using packet_id_t returning version? I suspect very very few.

> and natural interface for other MQTT library users.

I do not see how this is true.

Other libraries return the packet_id that they use because they are (mostly) not async libraries. 

- Libmosquitto, has own event loop - not async. Horrible to use. Absolutely awful. Incompatible with Boost.Asio, or really any other framework unless you run each framework in it's own thread. https://mosquitto.org/api/files/mosquitto-h.html#mosquitto_publish <- Is blocking call. Only connect can be called async.

- https://github.com/mqttjs/MQTT.js/blob/master/lib/client.js#L415 -- returns "this", not packet_id. Also, this code is not asynchronous.

```js
function sendPacket (client, packet, cb) {
  client.emit('packetsend', packet)

  var result = mqttPacket.writeToStream(packet, client.stream, client.options)

  if (!result && cb) {
    client.stream.once('drain', cb)
  } else if (cb) {
    cb()
  }
}
```

`mqttPacket.writeToStream(packet, client.stream, client.options)`  is not passed callback, to be called when write finishes. The callback is called synchronously. 

- https://github.com/eclipse/paho.mqtt.java/blob/master/org.eclipse.paho.client.mqttv3/src/main/java/org/eclipse/paho/client/mqttv3/IMqttAsyncClient.java#L364 -- both async and non-async, does not return packet_id_t, returns IMqttDeliveryToken (maybe this has packet_id_t inside?).

Those three examples are in C, JavaScript, and Java, respectively, and do not represent good practices when using C++ and Boost.Asio. CompletionToken method of doing Async networking is the right way in C++, and is not compatible with the function returning a packet_id_t.

Reply:
> I do not understand your reply.
> 
> I'm not proposing removing or moving acquire_unique_packet_id() in this change.

It seems that I'm misunderstanding something about your comment https://github.com/redboltz/mqtt_cpp/pull/547#discussion_r343400977. I'm sorry about my last reply. It was something rude. I will read carefully your comments again. So far, I don't understand yet. Maybe I need more time to understand the change correctly. In order to do that, we need to share not only the change in the PR but also future vision. Because some part of your update aims the future model. (Not mean future/promise 's future. In general meaning "future")
Maybe your last comment help that. Please wait a little more.

Reply:
My vision

```cpp
packet_id_t acquire_unique_packet_id();
void register_packet_id(packet_id_t);

// Non-async publish is complicated, because of qos::at_least_once, and storage consideration.
// If end user knows ahead of time that they are giving us qos::at_most_once, they can skip lifetime management entirely. However, that is not always something they know.
void publish
(
    packet_id_t pkt, // Pkt id established however the end-user wants. Mqtt_cpp just uses as provided.
    publish_options pubopt, // see https://github.com/redboltz/mqtt_cpp/pull/548
    as::const_buffer topic, // Lifetime managed by CompletionToken
    as::const_buffer content, // Lifetime managed by CompletionToken
    v5::properties,
    MQTT_NS::any lifetime
);

// Calls acquire_unique_packet_id(), then publish(packet_id_t, ...);
// For end user convenience. 
// Personally I don't think we should bother providing this, but it's not a large maintenance cost, so it's fine.
void publish
(
    publish_options pubopt, // see https://github.com/redboltz/mqtt_cpp/pull/548
    as::const_buffer topic, // Lifetime managed by CompletionToken
    as::const_buffer content, // Lifetime managed by CompletionToken
    v5::properties,
    MQTT_NS::any lifetime
);

void subscribe
(
    packet_id_t pkt,
    subscribe_options subopt,
    Iterator begin, // Each item in range [begin, end) will be passed to as::buffer()
    Iterator end, // Lifetime of items not managed. Non-async packet write.
    v5::properties
);

// Calls acquire_unique_packet_id(), then subscribe(packet_id_t, ...);
// For end user convenience. 
// Personally I don't think we should bother providing this, but it's not a large maintenance cost, so it's fine.
void subscribe
(
    subscribe_options subopt,
    Iterator begin, // Each item in range [begin, end) will be passed to as::buffer()
    Iterator end, // Lifetime of items not managed. Non-async packet write.
    v5::properties
);

void unsubscribe
(
    packet_id_t pkt,
    Iterator begin, // Each item in range [begin, end) will be passed to as::buffer()
    Iterator end, // Lifetime of items not managed. Non-async packet write.
    v5::properties
);

// Calls acquire_unique_packet_id(), then unsubscribe(packet_id_t, ...);
// For end user convenience. 
// Personally I don't think we should bother providing this, but it's not a large maintenance cost, so it's fine.
void unsubscribe
(
    Iterator begin, // Each item in range [begin, end) will be passed to as::buffer()
    Iterator end, // Lifetime of items not managed. Non-async packet write.
    v5::properties
);

template<typename CompletionToken>
CompletionToken::return_type
async_acquire_unique_packet_id
(
    packet_id_t & pkt, // Destination for the acquired packet id
    CompletionToken tok // keeps lifetime of pkt, tells async operation how to notify of completion (e.g. coroutine, std::future, callback, so on.
);

template<typename CompletionToken>
CompletionToken::return_type async_register_packet_id(packet_id_t, CompletionToken tok);

template<typename CompletionToken>
CompletionToken::return_type
async_publish
(
    packet_id_t pkt, // Pkt id established however the end-user wants. Mqtt_cpp just uses as provided.
    publish_options pubopt, // see https://github.com/redboltz/mqtt_cpp/pull/548
    as::const_buffer topic, // Lifetime managed by CompletionToken
    as::const_buffer content, // Lifetime managed by CompletionToken
    v5::properties,
    CompletionToken tok
);

// Calls async_acquire_unique_packet_id(), then async_publish(packet_id_t, ...);
// Important to note that async_acquire_unique_packet_id() is an async operation.
// The completiontoken given to async_acquire_unique_packet_id() will need to call async_publish() when the packet_id is ready.
// For end user convenience. 
// Personally I don't think we should bother providing this, but it's not a large maintenance cost, so it's fine.template<typename CompletionToken>
CompletionToken::return_type
async_publish
(
    publish_options pubopt, // see https://github.com/redboltz/mqtt_cpp/pull/548
    as::const_buffer topic, // Lifetime managed by CompletionToken
    as::const_buffer content, // Lifetime managed by CompletionToken
    v5::properties,
    CompletionToken tok
)

template<typename Iterator, typename CompletionToken>
CompletionToken::return_type
async_subscribe
(
    packet_id_t pkt,
    subscribe_options subopt,
    Iterator begin, // Each item in range [begin, end) will be passed to as::buffer()
    Iterator end, // Lifetime of items in range [begin, end) managed by CompletionToken
    v5::properties,
   CompletionToken
);

// Calls async_acquire_unique_packet_id(), then async_subscribe(packet_id_t, ...);
// Important to note that async_acquire_unique_packet_id() is an async operation.
// The completiontoken given to async_acquire_unique_packet_id() will need to call async_publish() when the packet_id is ready.
template<typename Iterator, typename CompletionToken>
CompletionToken::return_type
async_subscribe
(
    subscribe_options subopt,
    Iterator begin, // Each item in range [begin, end) will be passed to as::buffer()
    Iterator end, // Lifetime of items in range [begin, end) managed by CompletionToken
    v5::properties,
   CompletionToken
);

template<typename Iterator, typename CompletionToken>
CompletionToken::return_type
async_unsubscribe
(
    packet_id_t pkt,
    Iterator begin, // Each item in range [begin, end) will be passed to as::buffer()
    Iterator end, // Lifetime of items in range [begin, end) managed by CompletionToken
    v5::properties,
   CompletionToken
);

// Calls async_acquire_unique_packet_id(), then async_unsubscribe(packet_id_t, ...);
// Important to note that async_acquire_unique_packet_id() is an async operation.
// The completiontoken given to async_acquire_unique_packet_id() will need to call async_publish() when the packet_id is ready.
template<typename Iterator, typename CompletionToken>
CompletionToken::return_type
async_unsubscribe
(
    Iterator begin, // Each item in range [begin, end) will be passed to as::buffer()
    Iterator end, // Lifetime of items in range [begin, end) managed by CompletionToken
    v5::properties,
   CompletionToken
);
```

In this way, we provide

* publish
* subscribe
* unsubscribe
* async_publish
* async_subscribe
* async_unsubscribe

With *only* 12 functions total (6 provided for convenience)

This is because https://github.com/redboltz/mqtt_cpp/pull/548 removes publish_dup related functions
https://github.com/redboltz/mqtt_cpp/pull/547 (this PR) removes acquired_*** functions
https://github.com/redboltz/mqtt_cpp/issues/523 changes lifetime handling, so that end-user is responsible for keeping lifetime (manually, or by using CompletionToken)
and https://github.com/redboltz/mqtt_cpp/issues/536 (or something like it) removes all of the overloads where we have to manage different forms of strings / buffers.

Possibly keep version of (async_)subscribe / (async_)unsubscribe that take one value, for end user convenience. 


------------------------------------------------- 

https://github.com/redboltz/mqtt_cpp/pull/548 removes approximately 1000 lines of code, and makes it easier for the end user to control whether the published packet is a duplicate or not.

https://github.com/redboltz/mqtt_cpp/pull/547 (this PR) removes approximately 2000 lines of code, and makes it easier for the end user to control the packet_id, while reducing end user confusion about which functions to use.

https://github.com/redboltz/mqtt_cpp/issues/523 + https://github.com/redboltz/mqtt_cpp/issues/536  would remove approximately 1500

Reply:
See https://gist.github.com/inetic/dc9081baf45ec4b60037 for example of working with CompletionToken.

Reply:
The function would return: BOOST_ASIO_INITFN_RESULT_TYPE(CompletionToken, Signature), where the signature for async_publish, async_subscribe, and async_unsubscribe would be void(boost::system::error_code).

See bottom of this page: https://www.boost.org/doc/libs/1_67_0/boost/asio/async_result.hpp

Reply:
And finally, a much more in depth tutorial: https://www.boost.org/doc/libs/1_69_0/libs/beast/doc/html/beast/using_io/writing_composed_operations.html

Reply:
Last one: A description of how to adapt the boost.beast tutorial to generic boost.asio: https://www.reddit.com/r/cpp/comments/6aygos/boostasio_asynchronous_composed_operation_tutorial/

Reply:
I need to know very basic of your async API design.
To make things simple, I wrote an example.

I guess that `my_lib::async_do_some` interface and implementation is far from your vision.
https://wandbox.org/permlink/mA5lZ9iA0ihYpORX

How do you update it? and why?

I think that the current implementation is not bad, so far. It is a good start point to share the vision.

Reply:
I update exisiting library return type https://wandbox.org/permlink/nGBQV2776tsuSwQh

Reply:
This is the 2nd step. I guess that the return type should be something DEDUCED. https://www.boost.org/doc/libs/1_71_0/doc/html/boost_asio/reference/post.html
bool return type avoids that. 

So updated as follows:
https://wandbox.org/permlink/QgtUxAaho7EimeLb

Reply:
> * https://github.com/eclipse/paho.mqtt.java/blob/master/org.eclipse.paho.client.mqttv3/src/main/java/org/eclipse/paho/client/mqttv3/IMqttAsyncClient.java#L364 -- both async and non-async, does not return packet_id_t, returns IMqttDeliveryToken (maybe this has packet_id_t inside?), but IMqttDeliveryToken is also passed to async-callback, if used. So the returned IMqttDeliveryToken is only used in sync-version, but I see no explanation of why that is useful.

> maybe this has packet_id_t inside?

Yes, exactly. Paho calles packet_id (like structure) as token.

> So the returned IMqttDeliveryToken is only used in sync-version.

I don't think so. Even if the client is IMqttAsyncClient, return value IMqttDeliveryToken is used.

The publish callback is called when publish is finished. It is usually too late to know the packet_id.
And paho doesn't provides passing packet_id (IMqttDeliveryToken ) by caller version.

Reply:
I read your links. 

Let me explain what I studied.

I updated my example from https://wandbox.org/permlink/QgtUxAaho7EimeLb to https://wandbox.org/permlink/eMsByjHTfFJkN5al. The point is replacing `post()` with `async_wait()`. Because `post()` doesn't have error_code parameter.
In addition, return type become BOOST_ASIO_INITFN_RESULT_TYPE.
I think that my_lib can support callback, future, and async/await now.

`check_param()` is sync and private function. Even if the API is async, `check_param()` should be sync.
So the exception approach is reasonable.

Let's move on the next step.
I replaced `check_param()` with `create_id()`. The big difference is the result is required by caller.

I thought that if sync operation that returns value is private, then return the pair of 
the sync operation return value and BOOST_ASIO_INITFN_RESULT_TYPE(...).

This is updated code. Unfortunately it doesn't compile because `std::pair<int, void>` is not allowed. 
https://wandbox.org/permlink/UZKhAIbHuf2q2SSm

I can pass `int` as **out parameter** but it is not good because it requires breaking change. Even if I could return `std::pair<int, void>`, actually it is not possible, is also requires breaking change.

So if  I want to support CompletionToken pattern, the function shouldn't return any values. It should be reserved for `BOOST_ASIO_INITFN_RESULT_TYPE(...)`.

Fortunately, `mqtt_cpp`'s `acquire_unique_packet_id()` is public, so users can call it.

I updated the example.
https://wandbox.org/permlink/CmmktV8RRFwI6B6b

------

Now, I agree with removing internal `acquire_unique_packet_id()` call only from `async_*()` APIs. I agree with you **only this point**, so far.

------

I also read your vision. Thanks. I think that removing `std::string` version is too much. If we do that, we should provide some other casual way instead of `std::string` version. But this is not directly related this PR. Let's discuss it later.

------

You said that you are not going to remove **sync** APIs that returns packet_id and calls `acquire_unique_packet_id()` internally. It is very important point. The existing **sync** API call in examples and tests shouldn't be changed. For example 
https://github.com/redboltz/mqtt_cpp/pull/547/files#diff-62d04e1417cfcb95fd1edf0d04871f94R42

When you revert all of these kind of API calls, then I will merge the PR.
Of course calling `async_*()` calls in examples(may be no examples) and tests can be updated.

------

Reply:
I would rather merge this change first.

Can you review?

https://github.com/redboltz/mqtt_cpp/pull/548